### PR TITLE
[PERFSCALE-3438] Introduce UDN Density L3 workload

### PR DESCRIPTION
**Pull request by krishvoor**
This PR adds the newest workload to our kube-burner-ocp repo
as udn-density-l3-pods

https://github.com/kube-burner/kube-burner-ocp/pull/106

### CODE REVIEW

**Comment by josecastillolema** [comment]
> We need a new version/branch to be released on kube-burner-ocp and updates to e2e-benchmarking.

We also need to wait for the TP enablement step that @jtaleric has been working on @krishvoor in order to be able to test this.

**Comment by krishvoor** [comment]
> > We need a new version/branch to be released on kube-burner-ocp and updates to e2e-benchmarking.
> 
> We also need to wait for the TP enablement step that @jtaleric has been working on @krishvoor in order to be able to test this.

Currently, I'm focused on adding/introducing the workload through the PR.

For self-managed clusters, we can enable `FEATURE_GATES` which is an environment variable part of this workflow
https://steps.ci.openshift.org/workflow/openshift-qe-installer-aws#environment
And tackle it in a different PR.

**Comment by jtaleric** [comment]
Digging through the current ipi-conf-aws it appears introducing `FEATURE_GATE` at install would be good to get in before this, to ensure we can test this PR prior to merge, wdyt?

**Comment by krishvoor** [comment]
> Digging through the current ipi-conf-aws it appears introducing `FEATURE_GATE` at install would be good to get in before this, to ensure we can test this PR prior to merge, wdyt?

I will try it out on 4.18 OCP w/ `featuregate: TechPreviewNoUpgrade` and see how this goes through.

**Comment by krishvoor** [comment]
> > > We need a new version/branch to be released on kube-burner-ocp and updates to e2e-benchmarking.
> > 
> > 
> > We also need to wait for the TP enablement step that @jtaleric has been working on @krishvoor in order to be able to test this.
> 
> Currently, I'm focused on adding/introducing the workload through the PR.
> 
> For self-managed clusters, we can enable `FEATURE_GATES` which is an environment variable part of this workflow https://steps.ci.openshift.org/workflow/openshift-qe-installer-aws#environment And tackle it in a different PR.

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
Checked in one of the worker-nodes, the TechPreviewNoUpgrade is enabled.
```
cat /etc/kubernetes/kubelet.conf | grep TechPreviewNoUpgrade
  TechPreviewNoUpgrade: true
sh-5.1# exit
exit
sh-5.1# exit
exit

Removing debug pod ...
$ oc version
Client Version: 4.12.0-rc.6
Kustomize Version: v4.5.7
Kubernetes Version: v1.30.3
$
```

**Comment by krishvoor** [comment]
For some weird reason, the infra nodes are not coming online - needs investigation.

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Review by venkataanil — APPROVED** [review]
LGTM

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
Needs this: https://github.com/cloud-bulldozer/e2e-benchmarking/pull/722

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by jtaleric** [comment]
@krishvoor why can't we create infra nodes with this?

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+pushd e2e-benchmarking/workloads/kube-burner-ocp-wrapper
+export WORKLOAD=udn-density-l3-pods
+
+# current_worker_count=$(oc get nodes --no-headers -l node-role.kubernetes.io/worker=,node-role.kubernetes.io/infra!=,node-role.kubernetes.io/workload!= --output jsonpath="{.items[?(@.status.conditions[-1].type=='Ready')].status.conditions[-1].type}" | wc -w | xargs)
+
+# Run a non-indexed warmup for scheduling inconsistencies
+# ES_SERVER="" ITERATIONS=${current_worker_count} CHURN=false ./run.sh
+
+# The measurable run
+export ITERATIONS=50
```
Comment (jtaleric) [inline ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:37]: You should be able to remove this line.

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+fi
+
+REPO_URL="https://github.com/cloud-bulldozer/e2e-benchmarking";
+LATEST_TAG=$(curl -s "https://api.github.com/repos/cloud-bulldozer/e2e-benchmarking/releases/latest" | jq -r '.tag_name');
+TAG_OPTION="--branch $(if [ "$E2E_VERSION" == "default" ]; then echo "$LATEST_TAG"; else echo "$E2E_VERSION"; fi)";
+git clone $REPO_URL $TAG_OPTION --depth 1
+pushd e2e-benchmarking/workloads/kube-burner-ocp-wrapper
+export WORKLOAD=udn-density-l3-pods
+
+# current_worker_count=$(oc get nodes --no-headers -l node-role.kubernetes.io/worker=,node-role.kubernetes.io/infra!=,node-role.kubernetes.io/workload!= --output jsonpath="{.items[?(@.status.conditions[-1].type=='Ready')].status.conditions[-1].type}" | wc -w | xargs)
```
Comment (jtaleric) [inline ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:31]: Remove this.

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+# Run a non-indexed warmup for scheduling inconsistencies
+# ES_SERVER="" ITERATIONS=${current_worker_count} CHURN=false ./run.sh
```
Comment (jtaleric) [inline ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:34]: Remove

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+# The measurable run
+export ITERATIONS=50
+export CHURN=false
+
+export ES_SERVER="https://$ES_USERNAME:$ES_PASSWORD@$ES_HOST"
+
+if [[ "${ENABLE_LOCAL_INDEX}" == "true" ]]; then
+    EXTRA_FLAGS+=" --local-indexing"
+fi
+EXTRA_FLAGS+=" --iterations=50 --churn=false --gc-metrics=true --pod-ready-threshold=$POD_READY_THRESHOLD --profile-type=${PROFILE_TYPE}"
```
Comment (jtaleric) [inline ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:45]: Please set iterations to the env var that ref sets (which defaults to 50, but we should allow the job config to override)

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+export WORKLOAD=udn-density-l3-pods
+
+# current_worker_count=$(oc get nodes --no-headers -l node-role.kubernetes.io/worker=,node-role.kubernetes.io/infra!=,node-role.kubernetes.io/workload!= --output jsonpath="{.items[?(@.status.conditions[-1].type=='Ready')].status.conditions[-1].type}" | wc -w | xargs)
+
+# Run a non-indexed warmup for scheduling inconsistencies
+# ES_SERVER="" ITERATIONS=${current_worker_count} CHURN=false ./run.sh
+
+# The measurable run
+export ITERATIONS=50
+export CHURN=false
```
Comment (jtaleric) [inline ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:38]: Remove. this is referenced in the ref file.

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh (hunk 4 above)
Comment (jtaleric) [reply ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:45]: also use the var `${CHURM}` so the job can override.

**Comment by mohit-sheth** [comment]
@krishvoor is it possible to have 50 and 216 namespaces. 
50 is good for initial runs and it is super quick
216 will be our standard test once the feature is mature and can handle more UDNs. at this time we can retire 50 namespaces test
thoughts?
cc @jtaleric

File: ci-operator/config/openshift-qe/ocp-qe-perfscale-ci/openshift-qe-ocp-qe-perfscale-ci-main__aws-4.18-nightly-x86.yaml [hunk 6]
```diff
@@ -67,6 +67,26 @@ tests:
...
+  as: udn-density-l3-pods-24nodes
+  steps:
+    allow_skip_on_success: true
+    cluster_profile: aws-perfscale-qe
+    env:
+      BASE_DOMAIN: qe.devcluster.openshift.com
+      COMPUTE_NODE_REPLICAS: "24"
+      COMPUTE_NODE_TYPE: m5.xlarge
+      FEATURE_SET: TechPreviewNoUpgrade
+      IF_INSTALL_INFRA_WORKLOAD: "false"
```
Comment (jtaleric) [inline ci-operator/config/openshift-qe/ocp-qe-perfscale-ci/openshift-qe-ocp-qe-perfscale-ci-main__aws-4.18-nightly-x86.yaml:80]: I would like for us to keep infra nodes w/ ingress/monitoring.

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh (hunk 2 above)
Comment (jtaleric) [reply ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:31]: @mohit-sheth mentioned we should keep the multiplier of 9 and use the number of nodes to drive the iterations.

**Comment by jtaleric** [comment]
> @krishvoor is it possible to have 50 and 216 namespaces. 50 is good for initial runs and it is super quick 216 will be our standard test once the feature is mature and can handle more UDNs. at this time we can retire 50 namespaces test thoughts? cc @jtaleric

yeah, I like the idea of keeping a multiplier of 9.

File: ci-operator/config/openshift-qe/ocp-qe-perfscale-ci/openshift-qe-ocp-qe-perfscale-ci-main__aws-4.18-nightly-x86.yaml (hunk 6 above)
Comment (krishvoor) [reply ci-operator/config/openshift-qe/ocp-qe-perfscale-ci/openshift-qe-ocp-qe-perfscale-ci-main__aws-4.18-nightly-x86.yaml:80]: I hit issues similar to this. [Slack Thread](https://redhat-internal.slack.com/archives/C0780C0J6HJ/p1727900717063329?thread_ts=1727900311.516419&cid=C0780C0J6HJ) 
Thanks for tagging me on it :)

**Comment by krishvoor** [comment]
> > @krishvoor is it possible to have 50 and 216 namespaces. 50 is good for initial runs and it is super quick 216 will be our standard test once the feature is mature and can handle more UDNs. at this time we can retire 50 namespaces test thoughts? cc @jtaleric
> 
> yeah, I like the idea of keeping a multiplier of 9.

Ack, will bump it to 216 iterations for 24 nodes, i.e. 9 Iterations per worker node.
Also, I will drop a fix [here](https://github.com/cloud-bulldozer/e2e-benchmarking/blob/master/workloads/kube-burner-ocp-wrapper/run.sh#L132-L145)

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh (hunk 4 above)
Comment (krishvoor) [reply ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh:45]: It was getting bypassed here https://github.com/cloud-bulldozer/e2e-benchmarking/blob/master/workloads/kube-burner-ocp-wrapper/run.sh#L132-L145 for new workload, this was "hacky" to get things rolling.
Based on conversation, will drop a fix in e2e

**Comment by krishvoor** [comment]
> > > @krishvoor is it possible to have 50 and 216 namespaces. 50 is good for initial runs and it is super quick 216 will be our standard test once the feature is mature and can handle more UDNs. at this time we can retire 50 namespaces test thoughts? cc @jtaleric
> > 
> > 
> > yeah, I like the idea of keeping a multiplier of 9.
> 
> Ack, will bump it to 216 iterations for 24 nodes, i.e. 9 Iterations per worker node. Also, I will drop a fix [here](https://github.com/cloud-bulldozer/e2e-benchmarking/blob/master/workloads/kube-burner-ocp-wrapper/run.sh#L132-L145)

@jtaleric @mohit-sheth: Can you ptal?

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh (hunk 1 above)
Comment (krishvoor) [reply ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh]: Updated.

File: ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh (hunk 2 above)
Comment (krishvoor) [reply ci-operator/step-registry/openshift-qe/udn-density-l3-pods/openshift-qe-udn-density-l3-pods-commands.sh]: Ack, updated the iteration logic.

**Comment by krishvoor** [comment]
/pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
/pj-rehearse pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes

**Comment by krishvoor** [comment]
The changes here worked as expected: https://github.com/cloud-bulldozer/e2e-benchmarking/pull/724 
CI Job: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/57161/rehearse-57161-pull-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.18-nightly-x86-udn-density-l3-pods-24nodes/1841666847351508992

**Pull request merged**